core/src/main/java/com/linecorp/armeria/server/throttling/RequestThrottlingService.java

```diff
@@ -0,0 +1,81 @@
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.throttling;
+
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.server.HttpResponseException;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.SimpleDecoratingService;
+
+import io.micrometer.core.instrument.Tags;
+
+/**
+ * Decorates an HTTP {@link Service} to throttle incoming requests.
+ *
+ * @param <I> the {@link Request} type
+ * @param <O> the {@link Response} type
+ */
+public class RequestThrottlingService<I extends Request, O extends Response>
+        extends SimpleDecoratingService<I, O> {
+
+    /**
+     * Creates a new decorator using the specified {@link RequestThrottlingStrategy} instance.
+     *
+     * @param strategy The {@link RequestThrottlingStrategy} instance to be used
+     */
+    public static <I extends Request, O extends Response>
+    Function<Service<I, O>, RequestThrottlingService<I, O>>
+    newDecorator(RequestThrottlingStrategy strategy) {
+        return newDecorator((ctx, req) -> strategy);
+    }
+
+    /**
+     * Creates a new decorator with the specified {@link RequestThrottlingMapping}.
+     */
+    public static <I extends Request, O extends Response>
+    Function<Service<I, O>, RequestThrottlingService<I, O>> newDecorator(RequestThrottlingMapping mapping) {
+        return delegate -> new RequestThrottlingService<>(delegate, mapping);
+    }
+
+    private final RequestThrottlingMapping mapping;
+
+    /**
+     * Creates a new instance that decorates the specified {@link Service}.
+     */
+    protected RequestThrottlingService(Service<I, O> delegate,
+                                       RequestThrottlingMapping mapping) {
+        super(delegate);
+        this.mapping = mapping;
+    }
+
+    @Override
+    public O serve(ServiceRequestContext ctx, I req) throws Exception {
+        RequestThrottlingStrategy strategy = mapping.get(ctx, req);
+        if (strategy.shouldThrottle(req)) {
+            ctx.meterRegistry()
+               .counter("armeria_request_throttling", Tags.zip("name", strategy.name(),
+                                                               "type", "throttled_count"))
```

Comment:
```java
.counter(MeterRegistryUtil.name(registry, MeterUnit.NONE_TOTAL, "armeria", "server", "throttledRequests"),
         Tags.zip("name", strategy.name()))
```

Also, please keep the reference to the `Counter` so we don't keep looking it up. e.g. A weak map of strategy and `Counter`.